### ISSUE TITLE
[discussion] DevTools

### ISSUE THREAD

**Issue by jaredpalmer** [facebookexperimental/Recoil#194]
Not sure what people have in mind or are working on, but figured it would be good to centralize discussion.

**Comment by drarmstr** [comment]
Yup!  We've prototyped a little and have a project planned to develop.  But, would be interesting to hear what ideas the community has.  I'm working on tweaking the observability hook to improve support for debug tools.

**Comment by alexturpin** [comment]
I'm thinking a sort of tree view for all the state would be cool, if a convention for "namespace" atoms forms! So for instance if people start using `Module/Key` as "zones/contexts" as per this comment https://github.com/facebookexperimental/Recoil/issues/181#issuecomment-635003783, we'd be able to visualize state in a tree view where each `/` is a branch and as if state were a complex object instead of a flat list of keys internally.

Another idea would be being able to "watch" specific atoms in a watchlist for debugging things related to specific parts of state.

**Comment by jaredpalmer** [comment]
**re: entire tree**
- breakdwon by atoms, around naming convention `/`
- see diff on a given atom (use `deep-diff`)
- see atom history

btw the code in https://recoiljs.org/docs/guides/persistence#saving-state doesn't work. atomValues.get() doesn't return the value just metadata about persistence. it's hard to prototype without this working.

**Comment by drarmstr** [comment]
@jaredpalmer - were you looking at `atomInfo` instead of `atomValues`?

**Comment by smacpherson64** [comment]
Hrm, here when trying something similar `atomValues` seems to always be empty.

**Comment by drarmstr** [comment]
What I'm really looking forward to is a visualization of the data-flow graph of atoms, selectors, and subscribing components, with the ability to see current values and annotations/history on what has changed and how it has propagated through the graph.

**Comment by drarmstr** [comment]
@smacpherson64 `useTransactionObservation_UNSTABLE` is not a published API yet, the new API is under development and I hope to have it out soon.  Some things to keep in mind is that the current implementation was done for persistence, not debugging.  So, it only reports atoms values for atoms with special metadata marking them for persistence.  (The atom has a `persistence_UNSTABLE` property with a non-null `type` property.  It also doesn't report selectors.  Both issues will be addressed in the new API.

**Comment by smacpherson64** [comment]
> useTransactionObservation_UNSTABLE is not a published API yet, the new API is under development and I hope to have it out soon.

@drarmstr For sure! Please do not take this as "this thing isn't working"! I was using it for debugging to see things change. Just wanted to add to the conversation around `atomValues`! 

> So, it only reports atoms values for atoms with special metadata marking them for persistence. (The atom has a persistence_UNSTABLE property with a non-null type property. It also doesn't report selectors. Both issues will be addressed in the new API.

Makes sense, Recoil is powerful excited to watch it evolve!

**Comment by acutmore** [comment]
https://twitter.com/jaredpalmer/status/1266385084367732736?s=20 👀
This looks rad @jaredpalmer!

My Recoil devtools wishlist would be:

1. switch between different recoil roots
2. search/filter the list of atom/selectors by name
3. group history into the scheduler's transactions (so can see which values all changes together at the same time, verses which values changed separately) 
4. see number of active subscriptions for an atom/selector
5. a dropdown style tree of selector dependencies
6. manually update/reset/jump-back-to-a-previous-value of an atom
7. for the added code to expose these types of hooks to not add too much size to recoil in production

**Comment by ulises-jeremias** [comment]
Hi everyone! I'm working on this right now!

- [Recoil Devtools](https://www.npmjs.com/package/recoil-devtools)
- [Recoil Logger](https://www.npmjs.com/package/recoil-devtools-logger)

Both packages could be found in this [Github repository](https://github.com/ulises-jeremias/recoil-devtools), in which i'm creating devtools for recoil based on existing redux devtools and other projects. In future, more devtools and docs will be available!

**Comment by lokshunhung** [comment]
I used an empty redux store to mirror the changes the from recoil to redux devtool extension by [observing state changes in recoil](https://recoiljs.org/docs/guides/dev-tools#observing-all-state-changes).

Time traveling and probably most redux features does not work, but the redux devtool extension is integrated nicely into react native debugger so it's good enough for me.

If anyone is interested, here's a snippet:

```jsx
import { useEffect } from "react";
import { useRecoilSnapshot } from "recoil";
import { createStore } from "redux";

/**
 * Usage:
 * const App = () => (
 *   <RecoilRoot>
 *     <RecoilDebugObserver />
 *   </RecoilRoot>
 * );
 */
let RecoilDebugObserver = () => null;

const reduxDevtoolsExtension =
  typeof window !== "undefined" && window.__REDUX_DEVTOOLS_EXTENSION__;

if (reduxDevtoolsExtension) {
  const reducer = (state = {}, { type, payload }) => {
    if (String(type).startsWith("[recoil]")) {
      return { ...state, [payload.node.key]: payload.loadable };
    } else {
      return state;
    }
  };

  const store = createStore(
    reducer,
    reduxDevtoolsExtension({
      name: "recoil debug observer",
      maxAge: 100,
    })
  );

  RecoilDebugObserver = () => {
    const snapshot = useRecoilSnapshot();
    useEffect(() => {
      for (const node of snapshot.getNodes_UNSTABLE({ isModified: true })) {
        const loadable = snapshot.getLoadable(node);
        const action = {
          type: `[recoil] ${node.key}/${loadable.state}`,
          payload: { node, loadable },
        };
        store.dispatch(action);
      }
    }, [snapshot]);
    return null;
  };
}

export { RecoilDebugObserver };

```

**Comment by ulises-jeremias** [comment]
Hi just created this issue https://github.com/facebookexperimental/Recoil/issues/837 pressenting to the community some devtools I created for Recoil. Really similar to the redux devtools, so it should be easy to figure out how to use them for redux users too. A quick demo here: (see the issue for more context)

![devtools](https://user-images.githubusercontent.com/17727170/104832299-144ae980-586f-11eb-8958-8892ec990d10.gif)

**Comment by Vadorequest** [comment]
Using Next.js, what tool is available at this time? I've found https://github.com/open-source-labs/Recoilize, but it doesn't support Next.js yet.

@ulises-jeremias Are those tools mentioned in https://github.com/facebookexperimental/Recoil/issues/837 compatible with Next.js?

**Comment by ulises-jeremias** [comment]
@Vadorequest yes, as you can use it as any extra components lib on your project 👌🏻also, the development is really active, so any feedback or review is wellcome

**Comment by Vadorequest** [comment]
Thanks to @ulises-jeremias we've implemented Recoil Dev Tools with a Next.js project.

See PR at: https://github.com/Vadorequest/poc-nextjs-reaflow/pull/8
Demo: https://poc-nextjs-rea-git-fork-ulises-jeremias-main-ambroise-5619da.vercel.app/

It's the only solution I've found so far that works well with Next.js.

**Comment by LOGANLEEE** [comment]
any update for chrome extension?

**Comment by 72gm** [comment]
Hi @csantos42 , 

Any update on this? You mentioned you would be working on this in https://github.com/facebookexperimental/Recoil/issues/1233   

I'm trying to decide whether to use this in a production project I'm about to kick off so any timescales etc would be great

Thanks

**Comment by 72gm** [comment]
Hi @csantos42 , would be good if you could provide a roadmap for this? I'm sure a lot of people are interested in knowing

**Comment by creotip** [comment]
Hi guys, here you can check a package I created: 
https://recoil-gear.netlify.app/

It's a simple connector to redux devtools

**Comment by 72gm** [comment]
looks interesting, will have a look thanks @

disappointing that Facebook engineers can't be bothered to reply

**Comment by lokshunhung** [comment]
@72gm This is open source, there are no obligations that anyone must do anything.

If you feel that something is missing, you could contribute by submitting a PR or creating a new package, just like what @creotip did with [recoil-gear](https://github.com/creotip/recoil-gear/blob/38cf2340ae54c2901464732aa97a299a185bf7c8/libs/recoil-gear/src/lib/recoil-gear.tsx) with some modifications on [this](https://github.com/facebookexperimental/Recoil/issues/194#issuecomment-752375971).

The solution was present and you could already use it if you had searched a little harder. I guess maintainers and contributors would thank you if you could do some research and not vent your disappointment towards them, which just discourages people from contributing.

**Comment by korompaiistvan** [comment]
came to reaffirm the usefulness of a visual dataflow graph between atoms and selectors (as a first version I think this would be useful on its own, even without the subscribed components portion)

**Comment by dakom** [comment]
I see there is a devtools in the repo itself, what's that about?
